ironfish/src/memPool/recentlyEvictedCache.ts

```diff
@@ -0,0 +1,218 @@
+      return false
+    }
+
+    // keep the items in the two priority queues consistently in sync
+    this.expireAtQueue.remove(stringHash)
+    this.evictionQueue.remove(stringHash)
+
+    return true
+  }
+
+  /**
+   * Pops the transaction with the highest fee rate from the cache.
+   *
+   * The eviction and insertion queues will still be in sync after calling this method.
+   *
+   * @returns the transaction hash that was evicted, or undefined if the cache is empty
+   */
+  private poll(): TransactionHash | undefined {
+    const toEvict = this.evictionQueue.poll()
+    if (!toEvict) {
+      return
+    }
+
+    const hashToRemove = toEvict.hash
+    this.remove(hashToRemove)
+
+    return hashToRemove
+  }
+
+  /**
+   * @returns the number of transactions in the cache
+   */
+  size(): number {
+    return this.expireAtQueue.size()
+  }
+
+  /**
+   *
+   * @returns the usage of the recently evicted cache
+   */
+  saturation(): number {
+    return this.size() / this.maxSize
+  }
+
+  /**
+   * Adds a new transaction to the recently evicted cache.
+   * If the cache is full, the transaction with the highest fee rate will be evicted.
+   *
+   * Note that the cache is resized after the transaction is added. Thus, if the cache is above capacity
+   * and the new transaction has the highest fee rate in the cache, it will be immediately be evicted.
+   *
+   * @param transactionHash The hash of the transaction to add
+   * @param feeRate the fee/byte rate of the transaction
+   * @param currentBlockSequence the current block sequence when the transaction was added
+   * @param maxAge The maximum duration, in number of blocks, the transaction can stay in the cache
+   *
+   * @returns true if the transaction was successfully added to the cache, false if it was already present
+   */
+  add(
+    transactionHash: TransactionHash,
+    feeRate: bigint,
+    currentBlockSequence: number,
+    maxAge: number,
+  ): boolean {
+    if (this.has(transactionHash.toString('hex'))) {
+      // add to metrics that a duplicate was attempted to be added
+      return false
+    }
+
+    this.evictionQueue.add({
+      hash: transactionHash,
+      feeRate,
+    })
+
+    this.expireAtQueue.add({
+      hash: transactionHash,
+      expireAtSequence: currentBlockSequence + maxAge,
+    })
+
+    // keep the cache under max capacity
+    while (this.size() > this.maxSize) {
+      this.poll()
+    }
+
+    this.updateMetrics()
+
+    return true
+  }
+
+  /**
+   * Checks if the cache contains a transaction with the given hash.
+   *
+   * @returns true if the hash exists in the cache
+   */
+  has(hash: string): boolean {
+    return this.expireAtQueue.has(hash)
+  }
+
+  /**
+   * Flushes the cache of any transactions that will expire after adding the given block sequence.
+   *
+   * @param maxSequence All transactions with an expiration sequence <= `maxSequence` will be flushed.
+   */
+  flush(maxSequence: number): void {
```

Comment:
maybe just call this `sequence` to avoid terminology confusion around max vs min sequence

Reply:
changed